### PR TITLE
Feat/30/attachments service document generator

### PR DESCRIPTION
**Pull request by ponomarovv**
https://military-manager.atlassian.net/jira/software/projects/NET097/boards/1/backlog?selectedIssue=NET097-30&text=30

It works, except of no fonts issue in docker. It is not completely fixed. It is fixed only using dockerfile, not document generator code.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -46,3 +46,6 @@ Desktop.ini
 # macOS junk
 .DS_Store
 .idea
+
+# documents
+*.pdf
```
Comment (YevheniiVovk) [inline .gitignore]: Empty line

File: DocumentGenerator/FontResolver.cs [hunk 2]
```diff
@@ -86,7 +86,7 @@ private byte[] LoadFontData(string name)
 		{
 			try
 			{
-				using (var stream = new FileStream($"{_rootPath}\\fonts\\{name}", FileMode.Open, FileAccess.Read))
+				using (var stream = new FileStream($@"{_rootPath}/fonts/{name}", FileMode.Open, FileAccess.Read))
```
Comment (YevheniiVovk) [inline DocumentGenerator/FontResolver.cs]: Lets use class Path and his extension methods

File: DocumentGenerator/PdfGenerator.cs [hunk 3]
```diff
@@ -87,7 +87,9 @@ private static double GetTextHeight(XGraphics gfx, XFont font, string text, doub
 
 		public void AddTitle(string text)
 		{
-			AddTextBlock(text, fontStyle: XFontStyle.Bold);
+			// todo remove comments
+			//AddTextBlock(text, fontStyle: XFontStyle.Bold);
```
Comment (YevheniiVovk) [inline DocumentGenerator/PdfGenerator.cs]: If to do remove, please remove before commit

File: DocumentGenerator/PdfGenerator.cs [hunk 4]
```diff
@@ -87,7 +87,9 @@ private static double GetTextHeight(XGraphics gfx, XFont font, string text, doub
 
 		public void AddTitle(string text)
 		{
-			AddTextBlock(text, fontStyle: XFontStyle.Bold);
```
Comment (YevheniiVovk) [inline DocumentGenerator/PdfGenerator.cs]: Lets move bold option as parameter for method and possibly wrap it in config class

File: DocumentGenerator/XmlToDocumentConverter.cs [hunk 5]
```diff
@@ -64,7 +64,7 @@ private DocumentParams GetDocumentParameters(XmlDocument xmlDoc)
 				Right = 20
 			};
 
-			var docNode = xmlDoc.FirstChild.Name == "documnet" ? xmlDoc.FirstChild : null;
+			var docNode = xmlDoc.FirstChild.Name == "document" ? xmlDoc.FirstChild : null;
```
Comment (YevheniiVovk) [inline DocumentGenerator/XmlToDocumentConverter.cs]: Please remove change if it do not change anything

File: DocumentGenerator/XmlToDocumentConverter.cs [hunk 6]
```diff
@@ -122,7 +122,7 @@ public string CreateDocument(DocumentType type, string path, IDocumentData data)
 			}
 
 			documentGenerator.SaveDocument();
-			return $"{docName}.pdf";
+			return $@"{docName}.pdf";
```
Comment (YevheniiVovk) [inline DocumentGenerator/XmlToDocumentConverter.cs]: What reason ?

File: MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs [hunk 7]
```diff
@@ -40,25 +39,44 @@ public FileStreamResult GetDocument([FromQuery] string name)
...
-            var docName = _documentGenerationService.GeneratePdfDocument($"{_webRootPath}\\{_documentExportFolder}", templateName, templateData, jsonData);
+            _documentGenerationService.ApplyFontResolver("/wwwroot"); //_webRootPath);
+            var docName = _documentGenerationService.GeneratePdfDocument($@"{_webRootPath}/{_documentExportFolder}",
+                templateName, templateData, jsonData);
 
             return $"https://{Request.Host}/api/attachments/find?name={docName}";
         }
     }
-}
+}
```
Comment (YevheniiVovk) [inline MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs:105]: Empty line

File: MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs [hunk 8]
```diff
@@ -40,25 +39,44 @@ public FileStreamResult GetDocument([FromQuery] string name)
...
+
+           
 
-            _documentGenerationService.ApplyFontResolver(_webRootPath);
-            var docName = _documentGenerationService.GeneratePdfDocument($"{_webRootPath}\\{_documentExportFolder}", templateName, templateData, jsonData);
+            _documentGenerationService.ApplyFontResolver("/wwwroot"); //_webRootPath);
+            var docName = _documentGenerationService.GeneratePdfDocument($@"{_webRootPath}/{_documentExportFolder}",
+                templateName, templateData, jsonData);
 
             return $"https://{Request.Host}/api/attachments/find?name={docName}";
```
Comment (YevheniiVovk) [inline MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs:102]: Please use extension methods to build url

File: MilitaryManager.Attachments.API/MilitaryManager.Attachments.API.csproj [hunk 9]
```diff
@@ -7,6 +7,10 @@
     <DockerComposeProjectPath>..\docker-compose.dcproj</DockerComposeProjectPath>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Content Remove="wwwroot\documents\FontResolver.cs" />
+  </ItemGroup>
+
```
Comment (YevheniiVovk) [inline MilitaryManager.Attachments.API/MilitaryManager.Attachments.API.csproj]: ?

File: MilitaryManager.Attachments.API/wwwroot/data/document_templates/template_01.xml [hunk 10]
```diff
@@ -16,4 +16,4 @@
 	<p fontStyle="bold">Розгляд питань порядку денного:</p>
 	<ret>10</ret>
 	<list data="scheduleResults" ret="10">[#].    [text];</list>
-</documnet>
\ No newline at end of file
+</document>
```
Comment (YevheniiVovk) [inline MilitaryManager.Attachments.API/wwwroot/data/document_templates/template_01.xml]: empty line and changes with no reason

File: MilitaryManager.Attachments.API/wwwroot/data/document_templates/template_02.json [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+					{
+						"name": "text",
+						"type": "text",
+						"value": ""
+					}
+				]
+			}
+		}
+	]
+}
```
Comment (YevheniiVovk) [inline MilitaryManager.Attachments.API/wwwroot/data/document_templates/template_02.json]: Empty line

File: MilitaryManager.sln [hunk 12]
```diff
@@ -13,7 +13,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MilitaryManager.Units.API",
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MilitaryManager.IdentityServer", "MilitaryManager.IdentityServer\MilitaryManager.IdentityServer.csproj", "{B7C73A12-9CEB-44B9-9E36-2158734F580D}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DocumentGenerator", "DocumentGenerator\DocumentGenerator.csproj", "{2A473D5B-590C-4E0A-83FB-B2FE72D5B6F4}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "DocumentGenerator", "DocumentGenerator\DocumentGenerator.csproj", "{2A473D5B-590C-4E0A-83FB-B2FE72D5B6F4}"
```
Comment (YevheniiVovk) [inline MilitaryManager.sln]: ?

File: DocumentGenerator/FontResolver.cs (hunk 2 above)
Comment (ponomarovv) [reply DocumentGenerator/FontResolver.cs]: fixed

File: DocumentGenerator/PdfGenerator.cs (hunk 3 above)
Comment (ponomarovv) [reply DocumentGenerator/PdfGenerator.cs]: fixed

File: DocumentGenerator/XmlToDocumentConverter.cs (hunk 5 above)
Comment (ponomarovv) [reply DocumentGenerator/XmlToDocumentConverter.cs]: "document" and "documnet" (last 3 letters). There are different words, and i think it should be fixed.

File: DocumentGenerator/XmlToDocumentConverter.cs (hunk 6 above)
Comment (ponomarovv) [reply DocumentGenerator/XmlToDocumentConverter.cs]: There was no reason. i was tired. fixed

File: MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs (hunk 8 above)
Comment (ponomarovv) [reply MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs:102]: fixed

File: MilitaryManager.Attachments.API/MilitaryManager.Attachments.API.csproj (hunk 9 above)
Comment (ponomarovv) [reply MilitaryManager.Attachments.API/MilitaryManager.Attachments.API.csproj]: some strange bug that copied this file to that folder.

File: MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs (hunk 7 above)
Comment (ponomarovv) [reply MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs:105]: fixed

File: MilitaryManager.Attachments.API/wwwroot/data/document_templates/template_01.xml (hunk 10 above)
Comment (ponomarovv) [reply MilitaryManager.Attachments.API/wwwroot/data/document_templates/template_01.xml]: fixed

File: MilitaryManager.Attachments.API/wwwroot/data/document_templates/template_02.json (hunk 11 above)
Comment (ponomarovv) [reply MilitaryManager.Attachments.API/wwwroot/data/document_templates/template_02.json]: fixed

File: MilitaryManager.sln (hunk 12 above)
Comment (ponomarovv) [reply MilitaryManager.sln]: So. I used vs2019 and vs2022. and something went wrong...

File: DocumentGenerator/XmlToDocumentConverter.cs [hunk 13]
```diff
@@ -1,130 +1,138 @@
...
-				nodeParser.ParseNode(documentGenerator, node);
-			}
-
-			documentGenerator.SaveDocument();
-			return $"{docName}.pdf";
-		}
-
-		#endregion
-	}
+    /// <summary>
```
Comment (Alex-knu) [inline DocumentGenerator/XmlToDocumentConverter.cs]: Undo changes

File: DocumentGenerator/Services/DocumentGenerationService.cs [hunk 14]
```diff
@@ -6,85 +6,79 @@
...
-		public string GenerateWordDocument(string exportPath, string templateName, string templateData, string jsonData)
-		{
-			return GenerateDocument(DocumentType.Doc, exportPath, templateName, templateData, jsonData);
-		}
-
-		#endregion
-	}
+    /// <summary>
+    /// Service for generating documents
+    /// </summary>
```
Comment (Alex-knu) [inline DocumentGenerator/Services/DocumentGenerationService.cs]: Undo changes

File: DocumentGenerator/FontResolver.cs [hunk 15]
```diff
@@ -4,105 +4,104 @@
...
-
-		}
-
-		public string DefaultFontName => "times new roman";
-
-		#endregion
-	}
+    /// <summary>
+    /// Contains logic for getting custom fonts for document
+    /// </summary>
```
Comment (Alex-knu) [inline DocumentGenerator/FontResolver.cs]: Undo changes

File: MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs [hunk 16]
```diff
@@ -16,22 +18,28 @@ public class AttachmentsController : ControllerBase
...
             _documentGenerationService = service;
             _webRootPath = hostingEnvironment.WebRootPath;
             _logger = logger;
             _documentExportFolder = "documents";
+            
+            _unitService = UnitService;
         }
 
         [HttpGet]
         [Route("find")]
```
Comment (Alex-knu) [inline MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs]: Need discuss

File: MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs [hunk 17]
```diff
@@ -40,25 +48,50 @@ public FileStreamResult GetDocument([FromQuery] string name)
...
             {
                 _logger.LogError(ex, $"Template for {templateName} document is not available");
             }
 
-            var jsonData = "{\"city\":\"Рівне\",\"street\":\"Соборна\",\"number\":\"115\",\"date\":\"10.10.2022\"}";
+            var unit = _unitService.GetUnitAsync(1).Result;
+            var obj = new { name = unit.Name };
+            var jsonData2 = JsonConvert.SerializeObject(obj);
+
+            var jsonData = @"{city:'Рівне',currentDate:'08.09.2022',decreeNumber:'777'
```
Comment (Alex-knu) [inline MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs]: remove test data

File: MilitaryManager.Core/Interfaces/Services/IUnitService.cs [hunk 18]
```diff
@@ -10,5 +10,7 @@ public interface IUnitService
         Task<UnitDTO> CreateUnitAsync(UnitDTO query);
         Task<UnitDTO> UpdateUnitAsync(UnitDTO query);
         Task<UnitDTO> DeleteUnitAsync(int id);
+
```
Comment (Alex-knu) [inline MilitaryManager.Core/Interfaces/Services/IUnitService.cs:13]: Need discuss

File: MilitaryManager.Core/Services/UnitService.cs [hunk 19]
```diff
@@ -38,6 +38,15 @@ public async Task<IEnumerable<UnitDTO>> GetUnitsTreeAsync(int? id)
             return _mapper.Map<IEnumerable<UnitDTO>>(unitsTree);
         }
 
+
```
Comment (Alex-knu) [inline MilitaryManager.Core/Services/UnitService.cs:41]: Need discuss

File: MilitaryManager.Infrastructure/Data/Repositories/BaseRepository.cs [hunk 20]
```diff
@@ -78,6 +78,11 @@ public async Task<IEnumerable<TEntity>> GetListBySpecAsync(ISpecification<TEntit
             return await ApplySpecification(specification).ToListAsync();
         }
 
+        public async Task<TEntity> GetUnitBySpecAsync(ISpecification<TEntity> specification)
```
Comment (Alex-knu) [inline MilitaryManager.Infrastructure/Data/Repositories/BaseRepository.cs]: Need discuss

File: MilitaryManager.Units.API/appsettings.Development.json [hunk 21]
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "DefaultConnection": "Data Source=BAGUTA;Initial Catalog=Units;User ID=sa;Password=password1;Connect Timeout=30;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False"
+    "DefaultConnection": "Data Source=LPC;Initial Catalog=Units;User ID=sa;Password=password1;Connect Timeout=30;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False"
```
Comment (Alex-knu) [inline MilitaryManager.Units.API/appsettings.Development.json]: undo

File: MilitaryManager/client-app/.vscode/extensions.json [hunk 22]
```diff
@@ -1,4 +1,7 @@
 {
   // For more information, visit: https://go.microsoft.com/fwlink/?linkid=827846
```
Comment (Alex-knu) [inline MilitaryManager/client-app/.vscode/extensions.json:2]: ?

File: docker-compose.override.yml [hunk 23]
```diff
@@ -39,7 +40,7 @@ services:
     environment:
       - ASPNETCORE_ENVIRONMENT=Development
       - ASPNETCORE_URLS=https://+:443;http://+:80
-      - ConnectionStrings__DefaultConnection=Server=host.docker.internal;Database=IdentityServer_AspNet_Identity;User ID=Gwa_System;Password=T&t137S5tP!+3SMJQ<r0KJ;MultipleActiveResultSets=true;Application Name=IdentityServer
+      - ConnectionStrings__DefaultConnection=Server=host.docker.internal;Database=IdentityServer_AspNet_Identity;User ID=sa;Password=password1;MultipleActiveResultSets=true;Application Name=IdentityServer
```
Comment (Alex-knu) [inline docker-compose.override.yml]: undo

File: docker-compose.override.yml [hunk 24]
```diff
@@ -27,7 +28,7 @@ services:
       - ASPNETCORE_ENVIRONMENT=Development
       - ASPNETCORE_URLS=https://+:443;http://+:80
       - identityUrl=http://militarymanager.identity.server
-      - ConnectionStrings__DefaultConnection=Server=host.docker.internal;Database=Units;User ID=sa;Password=password1;MultipleActiveResultSets=true;Application Name=MilitaryManager.Units.API
+      - ConnectionStrings__DefaultConnection=Server=host.docker.internal;Database=Units;User ID=sa;Password=;MultipleActiveResultSets=true;Application Name=MilitaryManager.Units.API
```
Comment (Alex-knu) [inline docker-compose.override.yml]: undo

**Review by Alex-knu — CHANGES_REQUESTED** [review]

File: DocumentGenerator/Services/DocumentGenerationService.cs (hunk 14 above)
Comment (ponomarovv) [reply DocumentGenerator/Services/DocumentGenerationService.cs]: Ok, I think I can do it. but. on the left side we can see lines with first 8 spaces. on the right side we can see 4 leading spaces in the line. I think 4 spaces better.

File: DocumentGenerator/FontResolver.cs (hunk 15 above)
Comment (ponomarovv) [reply DocumentGenerator/FontResolver.cs]: Ok, I think I can do it. but. on the left side we can see lines with first 8 spaces. on the right side we can see 4 leading spaces in the line. I think 4 spaces better.

Comment (ponomarovv) [reply DocumentGenerator/FontResolver.cs]: We should discuss it with Vlad. Before doing it.

File: MilitaryManager.Units.API/appsettings.Development.json (hunk 21 above)
Comment (ponomarovv) [reply MilitaryManager.Units.API/appsettings.Development.json]: done

File: MilitaryManager/client-app/.vscode/extensions.json (hunk 22 above)
Comment (ponomarovv) [reply MilitaryManager/client-app/.vscode/extensions.json:2]: fixed with .gitignore

File: docker-compose.override.yml (hunk 23 above)
Comment (ponomarovv) [reply docker-compose.override.yml]: done

File: MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs (hunk 17 above)
Comment (ponomarovv) [reply MilitaryManager.Attachments.API/Controllers/AttachmentsController.cs]: hardcoded data were removed

File: MilitaryManager.Attachments.API/MilitaryManager.Attachments.API.csproj (hunk 9 above)
Comment (ponomarovv) [reply MilitaryManager.Attachments.API/MilitaryManager.Attachments.API.csproj]: fixed

File: docker-compose.override.yml (hunk 24 above)
Comment (ponomarovv) [reply docker-compose.override.yml]: done

File: MilitaryManager.Attachments.API/MilitaryManager.Attachments.API.csproj [hunk 25]
```diff
@@ -6,16 +6,19 @@
     <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
     <DockerComposeProjectPath>..\docker-compose.dcproj</DockerComposeProjectPath>
   </PropertyGroup>
-
-  <ItemGroup>
+	
+	<ItemGroup>
```
Comment (Alex-knu) [inline MilitaryManager.Attachments.API/MilitaryManager.Attachments.API.csproj]: ?

File: docker-compose.override.yml [hunk 26]
```diff
@@ -16,6 +16,7 @@ services:
     environment:
       - ASPNETCORE_ENVIRONMENT=Development
       - ASPNETCORE_URLS=https://+:443;http://+:80
+      - ConnectionStrings__DefaultConnection=Server=host.docker.internal;Database=Units;User ID=sa;Password=password1;MultipleActiveResultSets=true;Application Name=MilitaryManager.Attacjments.API
```
Comment (Alex-knu) [inline docker-compose.override.yml]: remove

**Review by Alex-knu — CHANGES_REQUESTED** [review]

File: docker-compose.override.yml (hunk 26 above)
Comment (ponomarovv) [reply docker-compose.override.yml]: it will be broken if we will remove this connection string.

**Review by Alex-knu — APPROVED** [review]

File: .gitignore [hunk 27]
```diff
@@ -46,3 +46,9 @@ Desktop.ini
...
```
Comment (vladtymo) [inline .gitignore]: Could be better to specify folder to ignore pdf files.

File: MilitaryManager.Infrastructure/Data/Repositories/BaseRepository.cs (hunk 20 above)
Comment (vladtymo) [reply MilitaryManager.Infrastructure/Data/Repositories/BaseRepository.cs]: Let's use an existing method.

**Review by vladtymo — COMMENTED** [review]
Merge the dev branch and fix all the conflicts.

**Review by vladtymo — APPROVED** [review]

File: .gitignore (hunk 27 above)
Comment (ponomarovv) [reply .gitignore]: done

**Pull request merged**